tests/validators/test_union.py

```diff
@@ -501,3 +501,267 @@ def test_left_to_right_union_strict():
+                    }
+                ),
+            ),
+            core_schema.model_schema(
+                ModelB, core_schema.model_fields_schema({'x': core_schema.model_field(core_schema.int_schema())})
+            ),
+        ]
+    )
+
+    validator = SchemaValidator(schema)
+
+    result = validator.validate_python({'x': 1})
+    assert isinstance(result, ModelA)
+    assert result.x == 1
+    assert result.y == 1
+
+    # passing a ModelB explicitly will not match the default value
+    b = ModelB()
+    assert validator.validate_python(b) is b
+
+
+def test_smart_union_model_field():
+    class ModelA:
+        x: int
+
+    class ModelB:
+        x: str
+
+    schema = core_schema.union_schema(
+        [
+            core_schema.model_schema(
+                ModelA, core_schema.model_fields_schema({'x': core_schema.model_field(core_schema.int_schema())})
+            ),
+            core_schema.model_schema(
+                ModelB, core_schema.model_fields_schema({'x': core_schema.model_field(core_schema.str_schema())})
+            ),
+        ]
+    )
+
+    validator = SchemaValidator(schema)
+
+    result = validator.validate_python({'x': 1})
+    assert isinstance(result, ModelA)
+    assert result.x == 1
+
+    result = validator.validate_python({'x': '1'})
+    assert isinstance(result, ModelB)
+    assert result.x == '1'
+
+
+def test_smart_union_dataclass_field():
+    @dataclass
+    class ModelA:
+        x: int
+
+    @dataclass
+    class ModelB:
+        x: str
+
+    schema = core_schema.union_schema(
+        [
+            core_schema.dataclass_schema(
+                ModelA,
+                core_schema.dataclass_args_schema(
+                    'ModelA', [core_schema.dataclass_field('x', core_schema.int_schema())]
+                ),
+                ['x'],
+            ),
+            core_schema.dataclass_schema(
+                ModelB,
+                core_schema.dataclass_args_schema(
+                    'ModelB', [core_schema.dataclass_field('x', core_schema.str_schema())]
+                ),
+                ['x'],
+            ),
+        ]
+    )
+
+    validator = SchemaValidator(schema)
+
+    result = validator.validate_python({'x': 1})
+    assert isinstance(result, ModelA)
+    assert result.x == 1
+
+    result = validator.validate_python({'x': '1'})
+    assert isinstance(result, ModelB)
+    assert result.x == '1'
+
+
+def test_smart_union_with_any():
+    """any is preferred over lax validations"""
+
+    # str not coerced to int
+    schema = core_schema.union_schema([core_schema.int_schema(), core_schema.any_schema()])
+    validator = SchemaValidator(schema)
+    assert validator.validate_python('1') == '1'
+
+    # int *is* coerced to float, this is a strict validation
+    schema = core_schema.union_schema([core_schema.float_schema(), core_schema.any_schema()])
+    validator = SchemaValidator(schema)
+    assert repr(validator.validate_python(1)) == '1.0'
```

Comment:
This is a known difference against main; on main this will not coerce and will keep as an `int`.

Reply:
> In particular, just want to make sure these behave the same as on main:

Specifically @dmontagu this one has changed, in discussion with @samuelcolvin we believe this to be okay.